main.go

```diff
@@ -0,0 +1,274 @@
+	msg := slackMessage{
+		Channel:  channel,
+		Text:     text,
+		Icon:     icon,
+		Username: user,
+	}
+
+	slackMsg, err := json.Marshal(msg)
+	if err != nil {
+		return nil, []string{}, err
+	}
+
+	return slackMsg, []string{channel}, nil
+}
+
+func (s *slackOutput) sendMessage(msg []byte, tag string) error {
+	lastError := fmt.Errorf("Unknown error in output")
+	for x := 0; x <= s.retryLimit; x++ {
+		// Don't send too quickly
+		s.rateLimiter.Wait(context.TODO())
+
+		// send the message to slack
+		resp, err := s.client.Post(s.slackURL, "application/json", bytes.NewReader(msg))
+		if err != nil {
+			if err, ok := err.(net.Error); ok && err.Timeout() {
+				// retry timeouts
+				lastError = err
+				continue
+			}
+			// Don't retry other errors
+			return fmt.Errorf("Failed to post to `%s`. Message: %s, Error: %s", tag, string(msg), err.Error())
+		}
+		defer resp.Body.Close()
+
+		// Attempt to get the response body to aid debugging
+		bodyString := ""
+		if resp.Body != nil {
+			bodyBytes, err := ioutil.ReadAll(resp.Body)
+			if err == nil {
+				bodyString = string(bodyBytes)
+			}
+		}
+
+		if resp.StatusCode == 200 {
+			// Done!
+			return nil
+		} else if resp.StatusCode == 429 {
+			// Rate limited. Retry after `Retry-After` seconds in the header
+			delayTime := rand.Intn(s.rateLimitConcurrency)
+			retry := resp.Header.Get("Retry-After")
+			if retry != "" {
+				sec, err := strconv.Atoi(retry)
+				if err != nil {
+					// Without a 'Retry-After' this will delay by the number of consumers
+					delayTime = delayTime + s.rateLimitConcurrency
+				} else {
+					delayTime = delayTime + sec
+				}
+			}
+			time.Sleep(time.Duration(delayTime) * time.Second)
+			lastError = fmt.Errorf("Failed to post to `%s`. Message: `%s`. Error: Exceeded rate limit", tag, string(msg))
+		} else if resp.StatusCode >= 500 && resp.StatusCode < 600 {
+			// A server side error occurred. Retry
+			lastError = fmt.Errorf("Failed to post to `%s`. Message: `%s`. Error: Server returned '%d': %s", tag, string(msg), resp.StatusCode, bodyString)
+		} else if resp.StatusCode >= 400 && resp.StatusCode < 500 {
+			// Don't retry 400s
+			return fmt.Errorf("Failed to post to `%s`. Message: `%s`. Error: Server returned '%d': %s", tag, string(msg), resp.StatusCode, bodyString)
+		}
+	}
+
+	// Retry attempts exceed
+	return fmt.Errorf("Retry limit `%d` exceeded posting to `%s`. Message: `%s`. Error: %s", s.retryLimit, tag, string(msg), lastError)
+}
+
+// ProcessMessage is called once per log to parse the log line and then reformat it
+// so that it can be directly used by the output. The returned tags will be passed along
+// with the encoded log to SendBatch()
+func (s *slackOutput) ProcessMessage(rawmsg []byte) (msg []byte, tags []string, err error) {
+	// Parse the log line
+	fields, err := decode.ParseAndEnhance(string(rawmsg), s.deployEnv, false, false, s.minTimestamp)
+	if err != nil {
+		return nil, []string{}, err
+	}
+
+	return s.encodeMessage(fields)
```

Comment:
Crazy idea: we could tag messages by username+icon.  That way the batcher could just concat all messages in a batch.  We'd also sent the max batch size to the max slack messages size.